File: benchmark/benchmark_sem_assertions_alinhamento_natural_v4/byte_level_memory.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>

#define MEM_SIZE 1024

/* ============================================================
 * Modelo Byte-Level
 * Memory : Addr -> Byte
 * ============================================================ */

typedef uint64_t Addr;
typedef uint8_t  Byte;

/* Memória */
Byte Memory[MEM_SIZE];

/* ============================================================
 * Escrita
 * ============================================================ */

void store8(Addr addr, Byte value)
{
    Memory[addr] = value;
}

/* ============================================================
 * Leituras
 * ============================================================ */

Byte load8(Addr addr)
{
    return Memory[addr];
}

/* Little-endian */
uint16_t load16(Addr addr)
{
    uint16_t b0 = Memory[addr];
    uint16_t b1 = Memory[addr + 1];

    return (b1 << 8) | b0;
}

/* Little-endian */
uint32_t load32(Addr addr)
{
    uint32_t b0 = Memory[addr];
    uint32_t b1 = Memory[addr + 1];
    uint32_t b2 = Memory[addr + 2];
    uint32_t b3 = Memory[addr + 3];

    return (b3 << 24) |
           (b2 << 16) |
           (b1 << 8)  |
            b0;
}

/* Little-endian */
uint64_t load64(Addr addr)
{
    uint64_t value = 0;

    for (int i = 0; i < 8; i++)
        value |= ((uint64_t)Memory[addr + i]) << (8 * i);

    return value;
}

/* ============================================================
 * Programa exemplo
 * ============================================================ */

int main(void)
{
    memset(Memory, 0, sizeof(Memory));

    Addr base = 100;

    /*
        char buf[4] = {0x11,0x22,0x33,0x44};
    */

    store8(base + 0, 0x11);
    store8(base + 1, 0x22);
    store8(base + 2, 0x33);
    store8(base + 3, 0x44);

    /*
        uint16_t *p = (uint16_t *)(buf + 1);

        Observe que este modelo NÃO verifica alinhamento.
        Apenas lê dois bytes consecutivos.
    */

    Addr p = base + 1;

    uint16_t x = load16(p);

    printf("Endereco base : %llu\n",
           (unsigned long long)base);

    printf("Endereco p    : %llu\n",
           (unsigned long long)p);

    printf("Valor lido    : 0x%04X\n", x);

    return 0;
}

File: benchmark/benchmark_sem_assertions_alinhamento_natural_v4/byte_level_memory.py
from z3 import *

# =====================================================
# Modelo de Memória Byte-Level
# =====================================================

# Endereços de 64 bits
AddrSort = BitVecSort(64)

# Cada posição da memória armazena 8 bits
ByteSort = BitVecSort(8)

# Memória: endereço -> byte
Memory = Array("Memory", AddrSort, ByteSort)

# =====================================================
# Operações de escrita
# =====================================================

def store8(mem, addr, value):
    """Escreve um byte."""
    return Store(mem, addr, value)


# =====================================================
# Operações de leitura
# =====================================================

def load8(mem, addr):
    """Lê um byte."""
    return Select(mem, addr)


def load16(mem, addr):
    """Lê 16 bits (little-endian)."""
    b0 = Select(mem, addr)
    b1 = Select(mem, addr + BitVecVal(1, 64))
    return Concat(b1, b0)


def load32(mem, addr):
    """Lê 32 bits (little-endian)."""
    b0 = Select(mem, addr)
    b1 = Select(mem, addr + BitVecVal(1, 64))
    b2 = Select(mem, addr + BitVecVal(2, 64))
    b3 = Select(mem, addr + BitVecVal(3, 64))
    return Concat(b3, b2, b1, b0)


def load64(mem, addr):
    """Lê 64 bits (little-endian)."""
    bytes = [
        Select(mem, addr + BitVecVal(i, 64))
        for i in range(8)
    ]
    return Concat(*reversed(bytes))


# =====================================================
# Programa exemplo
# =====================================================

solver = Solver()

mem = Memory

base = BitVecVal(1000, 64)

# Simula:
#
# char buf[4];
# buf = {0x11,0x22,0x33,0x44}
#

mem = store8(mem, base + 0, BitVecVal(0x11, 8))
mem = store8(mem, base + 1, BitVecVal(0x22, 8))
mem = store8(mem, base + 2, BitVecVal(0x33, 8))
mem = store8(mem, base + 3, BitVecVal(0x44, 8))

# Ponteiro desalinhado
#
# uint16_t *p = (uint16_t*)(buf + 1);
#

p = base + BitVecVal(1, 64)

# Leitura de 16 bits
x = load16(mem, p)

solver.add(
    x == BitVecVal(0x3322, 16)
)

# =====================================================
# Resultado
# =====================================================

print("Resultado:", solver.check())

if solver.check() == sat:
    print(solver.model())
